src/main/kotlin/io/meshcloud/dockerosb/persistence/ServiceInstanceRepository.kt

```diff
@@ -70,6 +82,72 @@ class ServiceInstanceRepository(private val yamlHandler: YamlHandler, private va
     return yamlHandler.readObject(instanceYml, ServiceInstance::class.java)
   }
 
+  // TODO consider introducing an common base type (interface) for all metric types,
+  // so we can get rid of ServiceInstanceDatapoints<*> and can use e.g. ServiceInstanceDatapoints<MetricModel>
+  fun tryGetServiceInstanceMetrics(serviceInstanceId: String, metricType: MetricType, from: Instant, to: Instant): List<ServiceInstanceDatapoints<*>>? {
+    val instanceMetricsYmlFiles = serviceInstanceMetricsYmlFiles(serviceInstanceId, metricType)
+    return when(metricType) {
+      MetricType.GAUGE -> {
+        var serviceInstanceDatapointsList: List<ServiceInstanceDatapoints<GaugeMetricModel>> = listOf()
+        val combinedInstanceMetricsList =  instanceMetricsYmlFiles.map{yamlHandler.readGeneric<ServiceInstanceDatapoints<GaugeMetricModel>>(it)}
+        var serviceInstanceIdGroup = combinedInstanceMetricsList.groupBy { it.serviceInstanceId }
+        for (uniqueServiceInstance in serviceInstanceIdGroup){
+          var resourceGroup = uniqueServiceInstance.value.groupBy { it.resource }
+          for (uniqueResource in resourceGroup ){
+            var filteredValues = uniqueResource.value
+                .flatMap { serviceInstanceDatapoints: ServiceInstanceDatapoints<GaugeMetricModel> -> serviceInstanceDatapoints.values }
+                .filterNot { gaugeMetricModel: GaugeMetricModel -> gaugeMetricModel.observedAt < from || gaugeMetricModel.observedAt > to }
+            if (filteredValues.count() > 0)
```

Comment:
t: We only need to know, that the result is not empty. Counting the elements is not required.
```suggestion
            if (filteredValues.isNotEmpty())
```

Same for the other `count() > 0` below.